### PR TITLE
[1/N] Core Data Models

### PR DESCRIPTION
**Pull request by Ritesh1905**
A split version of this [RFC PR](https://github.com/meta-pytorch/forge/pull/149)

### CODE REVIEW

File: src/forge/data_models/completion.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+from dataclasses import dataclass
+from typing import Optional
+
+import torch
+from forge.data_models.prompt import Prompt
+
+
+@dataclass
+class Completion:
```
Comment (pbontrager) [inline src/forge/data_models/completion.py:15]: Currently our Episode is a combination of Completion and ScoreCompletion and Experience. I think keeping them flat makes customization and logging a bit easier but either way can work.

File: src/forge/data_models/distributed_metric.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+
+import torch
+import torch.distributed as dist
+
+
+@dataclass
+class DistributedMetric(ABC):
```
Comment (pbontrager) [inline src/forge/data_models/distributed_metric.py:15]: I like this concept and it could be combined with our MetricLogger. But if we're going to handle this in a generic way, I don't think we want to deal with process groups, otherwise you have to set them up with a bunch of services that don't need them. Either we should use these abstractions and return metrics to a MetricService/controller to aggregate or handle it without formal abstractions, in a service specific way, as we do now.

File: src/forge/data_models/prompt.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+@dataclass
+class Message:
+    """A single message in a conversation."""
+
+    chunks: Sequence[str]
+    role: Role
+
+
+@dataclass
+class Prompt:
```
Comment (pbontrager) [inline src/forge/data_models/prompt.py:29]: In general we should match OpenAI requests here as the type. This way we can use the same formatting for local and API judge calls

File: src/forge/data_models/loss.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+from dataclasses import dataclass
+
+import torch
+from forge.data_models.distributed_metric import Fraction
+from forge.data_models.minibatch import Minibatch
+
+
+@dataclass
+class LossInput:
```
Comment (pbontrager) [inline src/forge/data_models/loss.py:15]: This is roughly what we're doing now, except it's trainer_logits + target_minibatch which is a subset of the minibatch you routed for the loss.
